src/main/kotlin/gg/ingot/iron/transformer/ValueTransformer.kt

```diff
@@ -1,155 +1,182 @@
+     * model, we'll send it to the model transformer to convert it.
      * @param resultSet The result set to retrieve the value from.
-     * @param columnName The column name to retrieve the value for.
-     * @param type The type of the collection.
-     */
-    private fun toCollection(resultSet: ResultSet, field: EntityField, namingStrategy: NamingStrategy): Collection<*> {
-        val arr = toArray(resultSet, field, namingStrategy) as Array<*>
-
-        val transformation = arrayTransformations.entries
-            // retrieve the first transformation that matches the type
-            .firstOrNull { it.key.java.isAssignableFrom(field.field.type) }
-            ?.value
-            ?: error("Unsupported collection type: ${field.field.type.name}")
-
-        return transformation(arr)
-    }
-
-    /**
-     * Retrieve the value as an object from the result set.
-     * @param resultSet The result set to retrieve the value from.
-     * @param field The field we're retrieving the value for.
-     * @param namingStrategy The naming strategy to use for the field.
+     * @param label The label to retrieve the value for.
+     * @param clazz The class to retrieve the value for.
      * @return The value from the result set.
      */
     @Suppress("UNCHECKED_CAST")
-    private fun toObject(resultSet: ResultSet, field: EntityField, namingStrategy: NamingStrategy): Any? {
-        val value = resultSet.getObject(field.convertedName(namingStrategy))
-
-        // Automatically map the enum to the enum type
-        if(field.isEnum && field.deserializer == null) {
-            return java.lang.Enum.valueOf(field.field.type as Class<out Enum<*>>, value as String)
+    fun read(resultSet: ResultSet, label: String?, clazz: Class<*>): Any? {
+        if (clazz.annotations.any { it is Model }) {
+            return iron.modelTransformer.readModel(resultSet, clazz)
         }
 
-        // Automatically convert Ints to Booleans for DBMS
-        // that don't give us back a boolean type.
-        if(value != null && value is Int && field.isBoolean) {
-            if(value != 0 && value != 1) {
-                error("Expected a boolean value, but found an integer value of $value for field: ${field.field.name}")
+        val value = if (label == null)
+            resultSet.getObject(1)
+        else if (clazz.isArray) resultSet.getArray(label).array
+        else resultSet.getObject(label)
+
+        if (value == null) return null
+
+        return if (clazz.isArray) {
+            value as Array<*>
+        } else if (clazz.isEnum) {
+            java.lang.Enum.valueOf(clazz as Class<out Enum<*>>, value as String)
+        } else if (Collection::class.java.isAssignableFrom(clazz)) {
+            (value as Array<*>).toList()
+        } else if (Set::class.java.isAssignableFrom(clazz)) {
+            (value as Array<*>).toSet()
+        } else if (box(clazz) == Boolean::class.java) {
+            when (value) {
```

Comment:
this entire when statement is one of the most disgusting and diabolical control flows i've ever seen in my life, fix this garbage immediately.

Reply:
Stale comment